tests/MongoDB.Driver.Core.Tests/Core/Compression/CompressedMessageBinaryEncoderTests.cs

```diff
@@ -0,0 +1,180 @@
+                    memoryStream.Position = 0;
+                    var result = subject.ReadMessage();
+                    result.ShouldBeEquivalentTo(
+                        compressedMessage,
+                        options =>
+                            options
+                                .Excluding(p => p.OriginalMessageStream));
+#else
+                    Record.Exception(() => { subject.WriteMessage(compressedMessage); })
+                        .Should().BeOfType<NotSupportedException>();
+#endif
+                }
+            }
+        }
+
+        private CommandResponseMessage CreateMessage(
+            int requestId = 0,
+            int responseTo = 0,
+            IEnumerable<CommandMessageSection> sections = null,
+            bool moreToCome = false)
+        {
+            sections = sections ?? new[] { CreateType0Section() };
+            return new CommandResponseMessage(new CommandMessage(requestId, responseTo, sections, moreToCome));
+        }
+
+        private CompressedMessageBinaryEncoder CreateSubject(Stream stream, params CompressorConfiguration[] compressors)
+        {
+            var compressorSource = GetCompressorSource(compressors);
+            var encoderSelector = new CommandResponseMessageEncoderSelector();
+            var subject = new CompressedMessageBinaryEncoder(stream, encoderSelector, compressorSource, _messageEncoderSettings);
+            return subject;
+        }
+
+        private Type0CommandMessageSection<RawBsonDocument> CreateType0Section(BsonDocument document = null)
+        {
+            document = document ?? new BsonDocument("t", 0);
+            var rawBson = new RawBsonDocument(document.ToBson());
+            return new Type0CommandMessageSection<RawBsonDocument>(rawBson, RawBsonDocumentSerializer.Instance);
+        }
+
+        private CompressedMessage GetCompressedMessage(MongoDBMessage message, CompressorType? compressorType = null)
+        {
+            if (!compressorType.HasValue)
+            {
+                compressorType = CompressorType.Zlib;
+            }
+
+            using (var memoryStream = new MemoryStream())
+            {
+                var encoderFactory = new BinaryMessageEncoderFactory(memoryStream, _messageEncoderSettings);
+                message.GetEncoder(encoderFactory).WriteMessage(message);
+                var byteBuffer = new ByteArrayBuffer(memoryStream.ToArray());
+                {
+                    var stream = new ByteBufferStream(byteBuffer);
+                    return new CompressedMessage(
+                        message,
+                        stream,
+                        compressorType.Value);
+                }
+            }
+        }
+
+        private ICompressorSource GetCompressorSource(params CompressorConfiguration[] compressors)
+        {
+            if (compressors == null)
+            {
+                return null;
+            }
+
+            if (!compressors.Any())
```

Comment:
As a general rule I recommend using LINQ mostly with IEnumerable. Some simple tests can be done more efficiently without using LINQ:

            if (compressors.Length == 0)

Reply:
p.s. I realize this is a test and this is not a performance critical line, but it's an observation worth making.